### PR TITLE
Add/hydra

### LINKED ISSUE #409: Polish up Hydra Code + Documentation + Examples

### PR DESCRIPTION
**Pull request by AndreBiedenkapp**
Fixes #409

### CODE REVIEW

File: smac/facade/hydra_facade.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        self.solver = None
+        self.rh = RunHistory(average_cost)
+        self.tae = tae(ta=self.scenario.ta, run_obj=self.scenario.run_obj)
+        self.tae_type = tae
+
+    def optimize(self):
+        """Optimizes the algorithm provided in scenario (given in constructor)
+
+        Returns
+        ----------
```
Comment (shukon) [inline smac/facade/hydra_facade.py]: too many '-'s

**Comment by AndreBiedenkapp** [comment]
@mlindauer This is ready to be reviewed again. I addressed all your feedback and worked on a very simple test that shows Hydras functionality without actually having to run any target algorithm.

File: doc/quickstart.rst [hunk 2]
```diff
@@ -383,3 +383,29 @@ The directory in which you invoked *SMAC* now contains a new folder called **SMA
...
+.. _hydra-example:
+
+Hydra on Spear-QCP
+------------------
+
+For this example we use *Hydra* to build a portfolio on the same example data presented in `Spear-QCP`__.
+Hydra is a portfolio builder that aims to build a portfolio by iteratively adding complementary configurations to the
+already existing portfolio. In the first iteration Hydra runs standard *SMAC* to determine a well performing configuration
+across all instances as a starting point for the portfolio. In following iterations Hydra only considers configurations
+that can improve the portfolio performance. Configurations that perform poor on instances which are already solved well
```
Comment (mlindauer) [inline doc/quickstart.rst]: Sorry, but the sentence is still not acceptable. Maybe we should replace the entire sentence?
I would recommend that we don't write what Hydra is *not* doing, but what it actually does. 
E.g., something along the following lines: "Hydra evaluates configurations in terms of performing better than the current portfolio."

File: smac/utils/io/cmd_reader.py [hunk 3]
```diff
@@ -131,6 +131,8 @@ def read_cmd(self):
         req_opts.add_argument("--random_configuration_chooser", default=None, type=FileType('r'),
                               help="[dev] path to a python module containing a class `RandomConfigurationChooserImpl`"
                                    "implementing the interface of `RandomConfigurationChooser`")
+        req_opts.add_argument("--hydra_iterations", default=3, type=int,
+                              help="[dev] number of hydra iterations")
```
Comment (mlindauer) [inline smac/utils/io/cmd_reader.py]: Please mention that this option is only active if Hydra mode is used.

File: smac/facade/hydra_facade.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        if self.output_dir is None:
+            self.top_dir = "hydra-output_%s" % (
+                datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d_%H:%M:%S_%f'))
+            self.scenario.output_dir = os.path.join(self.top_dir, "smac3-output_%s" % (
+                datetime.datetime.fromtimestamp(time.time()).strftime('%Y-%m-%d_%H:%M:%S_%f')))
+            self.output_dir = create_output_directory(self.scenario, run_id=self.run_id, logger=self.logger)
+
+        self.solver = SMAC(scenario=self.scenario, tae_runner=self.tae, **self.kwargs)
+        for i in range(self.n_iterations):
+            self.logger.info("Iteration: %d", (i + 1))
```
Comment (mlindauer) [inline smac/facade/hydra_facade.py]: Please change to "Hydra iteration:"

Comment (mlindauer) [reply smac/facade/hydra_facade.py]: And maybe add a special line before. e.g
self.logger.info("##########################################################")

**Comment by mlindauer** [comment]
And we could also print the entire portfolio in the end again or least write it in human-readable format into a file?

**Comment by shukon** [comment]
@mlindauer about the validation-warning: it is triggered, when reusing entries from runhistories, but validating with target algorithms (instead of EPM). It might be unnecessary altogether? I wanted it to serve as a disclaimer, but a warning is probably the wrong place, a warning in the docs might suffice.

**Comment by mlindauer** [comment]
>  a warning in the docs might suffice.

Yes.

**Pull request merged**